### ISSUE TITLE
Gibberish output while using average checkpoints for transducers

### ISSUE THREAD

**Issue by vinitunni** [espnet/espnet#4535]
**Describe the bug**
When using the average checkpoints option for RNN-Transducers, Inference stage gives gibberish results. THis was primarily noticed when trying to train our own models. But I am seeing this issue with PreTrained RNNT model too (librispeech pretrained RNNT model). However, to ensure there is no issue of dictionary or encoding, I am able to use pre-trained "conformer-sp" model to get proper outputs.
Because the problem is seen with pretrained models too, I am not sure how to fix/diagnose the issue on my end.

**Basic environments:**
- OS information: Linux 5.4.0-94-generic #106-Ubuntu SMP Thu Jan 6 23:58:14 UTC 2022 x86_64
- python version: `3.8.10 (default, Mar 15 2022, 12:22:08)  [GCC 9.4.0]`
- espnet version: `espnet 0.10.4`
- pytorch version: `pytorch 1.10.0+cu113`
 - Git hash [e.g. b88e89fc7246fed4c2842b55baba884fe1b4ecc2]   (Changes made to this base commit)
commit eee2566f2395d676e3978376315a4cd116a4e335
Author: Tomoki Hayashi <hayashi.tomoki@g.sp.m.is.nagoya-u.ac.jp>
Date:   Wed Nov 10 17:23:24 2021 +0900 

To Reproduce:
1) Run inference on pretrained RNNT models with averaging checkpoints

**Error logs**
On RNNT pretrained
'''
2022-07-26 08:39:54,657 (asr:1128) INFO: (1/271) decoding 1272-128104-0000
2022-07-26 08:40:07,244 (asr_utils:895) INFO: groundtruth: MISTER QUILTER IS THE APOSTLE OF THE MIDDLE CLASSES AND WE ARE GLAD TO WELCOME HIS GOSPEL
2022-07-26 08:40:07,244 (asr_utils:896) INFO: prediction : ▁A▁ASCENT▁OF▁THOSE▁HESSEXUASIVELUOUS▁A
2022-07-26 08:40:07,244 (asr:1128) INFO: (2/271) decoding 1272-128104-0001
2022-07-26 08:40:19,657 (asr_utils:895) INFO: groundtruth: NOR IS MISTER QUILTER'S MANNER LESS INTERESTING THAN HIS MATTER
2022-07-26 08:40:19,657 (asr_utils:896) INFO: prediction : ▁A▁ZEALOUSLY▁AS▁UM▁SOME
'''
On Conformer Pretrained
'''
2022-07-26 08:24:37,870 (asr:1128) INFO: (1/271) decoding 1272-128104-0000
2022-07-26 08:24:38,668 (e2e_asr_transformer:329) INFO: input lengths: 145
2022-07-26 08:24:38,668 (e2e_asr_transformer:345) INFO: max output length: 145
2022-07-26 08:24:38,668 (e2e_asr_transformer:346) INFO: min output length: 0
2022-07-26 08:26:19,425 (e2e_asr_transformer:475) INFO: end detected at 34
2022-07-26 08:26:19,439 (e2e_asr_transformer:508) INFO: total log probability: -1.961397159844637
2022-07-26 08:26:19,439 (e2e_asr_transformer:509) INFO: normalized log probability: -0.0817248816601932
2022-07-26 08:26:19,440 (asr_utils:895) INFO: groundtruth: MISTER QUILTER IS THE APOSTLE OF THE MIDDLE CLASSES AND WE ARE GLAD TO WELCOME HIS GOSPEL
2022-07-26 08:26:19,440 (asr_utils:896) INFO: prediction : ▁MISTER▁QUILTER▁IS▁THE▁APOSTLE▁OF▁THE▁MIDDLE▁CLASSES▁AND▁WE▁ARE▁GLAD▁TO▁WELCOME▁HIS▁GOSPEL<eos>
2022-07-26 08:26:19,440 (asr:1128) INFO: (2/271) decoding 1272-128104-0001
2022-07-26 08:26:21,845 (e2e_asr_transformer:329) INFO: input lengths: 119
2022-07-26 08:26:21,846 (e2e_asr_transformer:345) INFO: max output length: 119
2022-07-26 08:26:21,846 (e2e_asr_transformer:346) INFO: min output length: 0
2022-07-26 08:27:35,633 (e2e_asr_transformer:475) INFO: end detected at 19
2022-07-26 08:27:35,643 (e2e_asr_transformer:508) INFO: total log probability: -2.5117600932717323
2022-07-26 08:27:35,643 (e2e_asr_transformer:509) INFO: normalized log probability: -0.15698500582948327
2022-07-26 08:27:35,643 (asr_utils:895) INFO: groundtruth: NOR IS MISTER QUILTER'S MANNER LESS INTERESTING THAN HIS MATTER
2022-07-26 08:27:35,644 (asr_utils:896) INFO: prediction : ▁NOR▁IS▁MISTER▁QUILTER'S▁MANNER▁LESS▁INTERESTING▁THAN▁HIS▁MATTER<eos>

'''

**Comment by b-flo** [comment]
Hi,

> When using the average checkpoints option for RNN-Transducers, Inference stage gives gibberish results. THis was primarily noticed when trying to train our own models. But I am seeing this issue with PreTrained RNNT model too (librispeech pretrained RNNT model). However, to ensure there is no issue of dictionary or encoding, I am able to use pre-trained "conformer-sp" model to get proper outputs.
> Because the problem is seen with pretrained models too, I am not sure how to fix/diagnose the issue on my end.

Sorry, I'm not sure I'm following everything.

1. Are we talking about ESPnet1 or ESPnet2 here?
2. Which Librispeech pre-trained models are you using?
3. The issue only happens when using a model with average checkpoints but everything is fine when taking the best one?

**Comment by vinitunni** [comment]
Hello,

>    Are we talking about ESPnet1 or ESPnet2 here?

I am talking about ESPnet1 here

> Which Librispeech pre-trained models are you using?

I am using the models mentioned in the [Results page](https://github.com/espnet/espnet/blob/master/egs/librispeech/asr1/RESULTS.md)
For RNNT, "pytorch large conformer-transducer with specaug + speed perturbation (4 GPUs)"
For conformer, "pytorch large conformer with specaug + speed perturbation (8 GPUs) + Transformer LM (4 GPUs)"

> The issue only happens when using a model with average checkpoints but everything is fine when taking the best one?

Yes, wierdly, in the models I have trained, if I take an average of 1 or the best model (or even a single snapshot), inferences are fine. An average of more than one gives gibberish


> decoded with an average over last 1 checkpoint


 id: (2277_149897-2277-149897-0035)
Scores: (#C #S #D #I) 6 0 0 0
REF:  very truly yours et cetera compromise 
HYP:  very truly yours et cetera compromise 
Eval:                                       

id: (2277_149897-2277-149897-0036)
Scores: (#C #S #D #I) 16 0 0 0
REF:  so here it was spread out clear before him and now he knew what to expect 
HYP:  so here it was spread out clear before him and now he knew what to expect 
Eval:                                                                           

id: (2277_149897-2277-149897-0037)
Scores: (#C #S #D #I) 12 0 0 0
REF:  if he didn't go and see them they would sue him promptly 
HYP:  if he didn't go and see them they would sue him promptly 
Eval:                                                          
> decoded with an average over last 5 checkpoints

id: (2277_149897-2277-149897-0035)
Scores: (#C #S #D #I) 0 4 2 0
REF:  VERY TRULY YOURS ET   CETERA     COMPROMISE 
HYP:  **** ***** AND   RACE ALTOGETHER ADDED      
Eval: D    D     S     S    S          S          

id: (2277_149897-2277-149897-0036)
Scores: (#C #S #D #I) 0 1 15 0
REF:  SO HERE IT WAS SPREAD OUT CLEAR BEFORE HIM AND NOW HE KNEW WHAT TO EXPECT 
HYP:  ** **** ** *** ****** *** ***** ****** *** *** *** ** **** **** ** IRONTE 
Eval: D  D    D  D   D      D   D     D      D   D   D   D  D    D    D  S      

id: (2277_149897-2277-149897-0037)
Scores: (#C #S #D #I) 0 4 8 0
REF:  IF HE DIDN'T GO AND SEE THEM THEY WOULD  SUE      HIM    PROMPTLY 
HYP:  ** ** ****** ** *** *** **** **** ALMOST ADVANCED ALMOST ALWAYS

**Comment by b-flo** [comment]
Hum, it looks strangely familiar to what we observed in #3948 but the related bug is already fixed

Is your repo up-to-date? If the information you provided in "Basic environments:" are correct, you're missing the fix.

**Comment by vinitunni** [comment]
Yes,
The changes to 'blocks.py' from this [PR](https://github.com/espnet/espnet/pull/4020/files) has solved this issue.
Thank you.
